File: src/reader.ts
import fs from 'node:fs';
import path from 'node:path';
import { x } from 'tinyexec';

export interface ReadOptions {
  cwd?: string;
  edit?: string | boolean;
  env?: string;
  from?: string;
  to?: string;
  last?: boolean;
  fromLastTag?: boolean;
  gitLogArgs?: string;
}

export async function readCommitMessages(options: ReadOptions = {}): Promise<string[]> {
  const cwd = options.cwd || process.cwd();

  // 1. Read from specified file or default edit message
  if (options.edit || options.env) {
    let filePath: string | null = null;
    if (options.env && process.env[options.env]) {
      filePath = process.env[options.env]!;
    } else if (typeof options.edit === 'string' && options.edit.length > 0) {
      filePath = path.isAbsolute(options.edit) ? options.edit : path.join(cwd, options.edit);
    } else {
      filePath = path.join(cwd, '.git', 'COMMIT_EDITMSG');
    }

    if (filePath && fs.existsSync(filePath)) {
      const content = fs.readFileSync(filePath, 'utf8');
      // Strip comments (e.g. lines starting with '#')
      const cleaned = content
        .split(/\r?\n/)
        .filter((line) => !line.trim().startsWith('#'))
        .join('\n')
        .trim();
      return cleaned ? [cleaned] : [];
    }
  }

  // 2. Read git history using `git log`
  if (options.from || options.to || options.last || options.fromLastTag) {
    let from = options.from;
    const to = options.to || 'HEAD';

    if (options.last && !from) {
      from = 'HEAD~1';
    } else if (options.fromLastTag && !from) {
      try {
        const tagResult = await x('git', ['describe', '--tags', '--abbrev=0'], { nodeOptions: { cwd } });
        if (tagResult.exitCode === 0) {
          from = tagResult.stdout.trim();
        }
      } catch {
        from = 'HEAD~1';
      }
    }

    const args = ['log', '--format=%B%x1e'];
    if (from && to) {
      args.push(`${from}..${to}`);
    } else if (from) {
      args.push(`${from}..HEAD`);
    } else if (to) {
      args.push(to);
    }

    if (options.gitLogArgs) {
      args.push(...options.gitLogArgs.split(' ').filter(Boolean));
    }

    try {
      const result = await x('git', args, { nodeOptions: { cwd } });
      if (result.exitCode === 0) {
        return result.stdout
          .split('\x1e')
          .map((msg) => msg.trim())
          .filter(Boolean);
      }
    } catch {
      // Fallback
    }
  }

  return [];
}

export async function readStdin(): Promise<string> {
  if (process.stdin.isTTY) return '';
  process.stdin.setEncoding('utf8');
  let result = '';
  for await (const chunk of process.stdin) {
    result += chunk;
  }
  return result.trim();
}

File: src/reader.rs
use std::fs;
use std::io::{self, Read};
use std::process::Command;

#[derive(Default)]
pub struct ReadOptions {
    pub cwd: Option<String>,
    pub edit: Option<String>,
    pub env: Option<String>,
    pub from: Option<String>,
    pub to: Option<String>,
    pub last: bool,
    pub from_last_tag: bool,
    pub git_log_args: Option<String>,
}

pub fn read_commit_messages(options: &ReadOptions) -> Vec<String> {
    let cwd = options.cwd.as_deref().unwrap_or(".");

    if let Some(edit) = &options.edit {
        let path = if edit == "true" || edit.is_empty() {
            format!("{}/.git/COMMIT_EDITMSG", cwd)
        } else {
            edit.clone()
        };
        if let Ok(content) = fs::read_to_string(&path) {
            let cleaned = content
                .lines()
                .filter(|l| !l.trim().starts_with('#'))
                .collect::<Vec<&str>>()
                .join("\n")
                .trim()
                .to_string();
            if !cleaned.is_empty() {
                return vec![cleaned];
            }
        }
    }

    if options.from.is_some() || options.to.is_some() || options.last || options.from_last_tag {
        let mut from = options.from.clone();
        let to = options.to.as_deref().unwrap_or("HEAD");

        if options.last && from.is_none() {
            from = Some("HEAD~1".to_string());
        } else if options.from_last_tag && from.is_none() {
            if let Ok(out) = Command::new("git")
                .args(["describe", "--tags", "--abbrev=0"])
                .current_dir(cwd)
                .output()
            {
                if out.status.success() {
                    from = Some(String::from_utf8_lossy(&out.stdout).trim().to_string());
                }
            }
            if from.is_none() {
                from = Some("HEAD~1".to_string());
            }
        }

        let mut args = vec!["log".to_string(), "--format=%B%x1e".to_string()];
        if let Some(f) = &from {
            args.push(format!("{}..{}", f, to));
        }

        if let Ok(out) = Command::new("git").args(&args).current_dir(cwd).output() {
            if out.status.success() {
                let stdout = String::from_utf8_lossy(&out.stdout);
                return stdout
                    .split('\x1e')
                    .map(|s| s.trim().to_string())
                    .filter(|s| !s.is_empty())
                    .collect();
            }
        }
    }

    Vec::new()
}

pub fn read_stdin() -> String {
    let mut buffer = String::new();
    let _ = io::stdin().read_to_string(&mut buffer);
    buffer.trim().to_string()
}
